[BE][SyncBN] Avoid sync stats in eval mode

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56982 [BE][SyncBN] Avoid sync stats in eval mode**

SyncBatchNorm should behave as a regular BN layer in eval model, this
change ensures that this is the case.

In particular, the bug was when `track_running_stats=False`, `bn_training` would be set to True in eval mode, but this would trigger a collective sync in syncBN.

When `track_running_stats=True`, this is not an issue as in inference mode, `(self.running_mean is None) and (self.running_var is None)` would evaluate to False and we would not sync either way.

However, in eval mode syncBN should behave like a regular BN layer and not do this sync.

Closes https://github.com/pytorch/pytorch/issues/48988

Ensured with unittest that when used for inference on a single rank, stats sync is not triggered.

Differential Revision: [D27579297](https://our.internmc.facebook.com/intern/diff/D27579297/)

rohan-varma committed:
[BE][SyncBN] Avoid sync stats in eval mode

SyncBatchNorm should behave as a regular BN layer in eval model, this
change ensures that this is the case.

In particular, the bug was when `track_running_stats=False`, `bn_training` would be set to True in eval mode, but this would trigger a collective sync in syncBN.

However, in eval mode syncBN should behave like a regular BN layer and not do this sync.

Closes https://github.com/pytorch/pytorch/issues/48988

Ensured with unittest that when used for inference on a single rank, stats sync is not triggered.

Differential Revision: [D27579297](https://our.internmc.facebook.com/intern/diff/D27579297/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[BE][SyncBN] Avoid sync stats in eval mode"


SyncBatchNorm should behave as a regular BN layer in eval model, this
change ensures that this is the case.

In particular, the bug was when `track_running_stats=False`, `bn_training` would be set to True in eval mode, but this would trigger a collective sync in syncBN.

However, in eval mode syncBN should behave like a regular BN layer and not do this sync.

Closes https://github.com/pytorch/pytorch/issues/48988

Ensured with unittest that when used for inference on a single rank, stats sync is not triggered.

Differential Revision: [D27579297](https://our.internmc.facebook.com/intern/diff/D27579297/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[BE][SyncBN] Avoid sync stats in eval mode"


SyncBatchNorm should behave as a regular BN layer in eval model, this
change ensures that this is the case.

In particular, the bug was when `track_running_stats=False`, `bn_training` would be set to True in eval mode, but this would trigger a collective sync in syncBN.

However, in eval mode syncBN should behave like a regular BN layer and not do this sync.

Closes https://github.com/pytorch/pytorch/issues/48988

Ensured with unittest that when used for inference on a single rank, stats sync is not triggered.

Differential Revision: [D27579297](https://our.internmc.facebook.com/intern/diff/D27579297/)

[ghstack-poisoned]

wayi1 reviewed (approved):
I am surprised that the widely used `SyncBatchNorm` has this bug.

This PR should be able to give a nice improvement. Better to do some evaluation on some CV use cases.

rohan-varma committed:
Update on "[BE][SyncBN] Avoid sync stats in eval mode"


SyncBatchNorm should behave as a regular BN layer in eval model, this
change ensures that this is the case.

In particular, the bug was when `track_running_stats=False`, `bn_training` would be set to True in eval mode, but this would trigger a collective sync in syncBN.

When `track_running_stats=True`, this is not an issue as in inference mode, `(self.running_mean is None) and (self.running_var is None)` would evaluate to False and we would not sync either way.

However, in eval mode syncBN should behave like a regular BN layer and not do this sync.

Closes https://github.com/pytorch/pytorch/issues/48988

Ensured with unittest that when used for inference on a single rank, stats sync is not triggered.

Differential Revision: [D27579297](https://our.internmc.facebook.com/intern/diff/D27579297/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[BE][SyncBN] Avoid sync stats in eval mode"


SyncBatchNorm should behave as a regular BN layer in eval model, this
change ensures that this is the case.

In particular, the bug was when `track_running_stats=False`, `bn_training` would be set to True in eval mode, but this would trigger a collective sync in syncBN.

When `track_running_stats=True`, this is not an issue as in inference mode, `(self.running_mean is None) and (self.running_var is None)` would evaluate to False and we would not sync either way.

However, in eval mode syncBN should behave like a regular BN layer and not do this sync.

Closes https://github.com/pytorch/pytorch/issues/48988

Ensured with unittest that when used for inference on a single rank, stats sync is not triggered.

Differential Revision: [D27579297](https://our.internmc.facebook.com/intern/diff/D27579297/)

[ghstack-poisoned]

rohan-varma commented:
Failures are unrelated: FAIL [0.049s]: test_eig_with_eigvec_cuda_float64 (__main__.TestTensorDeviceOpsCUDA) (rocm) and ratelim issue.
